mongo/mongo.go

```diff
@@ -136,20 +149,64 @@ func transformBsoncoreDocument(registry *bsoncodec.Registry, val interface{}, ma
 		// Slight optimization so we'll just use MarshalBSON and not go through the codec machinery.
 		val = bson.Raw(bs)
 	}
-	if !mapAllowed {
-		refValue := reflect.ValueOf(val)
-		if refValue.Kind() == reflect.Map && refValue.Len() > 1 {
-			return nil, ErrMapForOrderedArgument{paramName}
-		}
+
+	buf := new(bytes.Buffer)
+	enc, err := getEncoder(buf, bsonOpts, registry)
+	if err != nil {
+		return nil, fmt.Errorf("error configuring BSON encoder: %w", err)
 	}
 
-	// TODO(skriptble): Use a pool of these instead.
-	buf := make([]byte, 0, 256)
-	b, err := bson.MarshalAppendWithRegistry(registry, buf[:0], val)
+	err = enc.Encode(val)
 	if err != nil {
 		return nil, MarshalError{Value: val, Err: err}
 	}
-	return b, nil
+
+	return buf.Bytes(), nil
+}
+
+// ensureID inserts the given ObjectID as an element named "_id" at the
+// beginning of the given BSON document if there is not an "_id" already. If
+// there is already an element named "_id", the document is not modified. It
+// returns the resulting document and the decoded Go value of the "_id" element.
+func ensureID(
+	doc bsoncore.Document,
+	oid primitive.ObjectID,
+	bsonOpts *options.BSONOptions,
+	reg *bsoncodec.Registry,
+) (bsoncore.Document, interface{}, error) {
+	if reg == nil {
+		reg = bson.DefaultRegistry
+	}
+
+	// Try to find the "_id" element. If it exists, try to unmarshal just the
+	// "_id" field as an interface{} and return it along with the unmodified
+	// BSON document.
+	if _, err := doc.LookupErr("_id"); err == nil {
+		var id struct {
+			ID interface{} `bson:"_id"`
+		}
+		dec, err := getDecoder(doc, bsonOpts, reg)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error configuring BSON decoder: %w", err)
+		}
+		err = dec.Decode(&id)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error unmarshaling BSON document: %w", err)
+		}
+
+		return doc, id.ID, nil
+	}
+
+	// If we couldn't find an "_id" element, add one with the value of the
+	// provided ObjectID.
+	olddoc := doc
+	doc = make(bsoncore.Document, 0, len(olddoc)+17) // type byte + _id + null byte + object ID
```

Comment:
Could we define a constant indicating where the "17" comes from?

Reply:
Will do.